src/xmipp/libraries/tomo/tomo_detect_misalignment_residuals.cpp

```diff
@@ -0,0 +1,843 @@
+	double mahaThr = avgMahaDist + 3 * stdMahaDist;
+
+	// Global alignment analysis
+	std::cout << "---------------- Global misalignemnt analysis" << std::endl;
+
+	double rationMisalignedChains = 0;	// 0 -> aligned -- 1 -> misaligned 
+
+	for (size_t n = 0; n < numberOfInputCoords; n++)
+	{
+		std::vector<resMod> resMod_fid;
+		getResModByFiducial(n, resMod_fid);
+
+		size_t numberResMod = resMod_fid.size();
+
+		sumMahaDist = 0;
+		sumMahaDist2 = 0;
+
+		for (size_t i = 0; i < numberResMod; i++)
+		{
+			if (removeOutliers && resMod_fid[i].mahalanobisDistance < mahaThr)
+			{		
+				sumMahaDist += resMod_fid[i].mahalanobisDistance;
+				sumMahaDist2 += resMod_fid[i].mahalanobisDistance * resMod_fid[i].mahalanobisDistance;
+			}
+		}
+
+		avgMahaDist = sumMahaDist / numberResMod;
+
+		if (avgMahaDist > 1)
+		{
+			rationMisalignedChains += 1;
+		}
+		
+		#ifdef DEBUG_RESIDUAL_ANALYSIS
+		std::cout << "Average mahalanobis distance for 3D cooridinate " << n << ": " << avgMahaDist << std::endl;
+		#endif
+	}
+
+	if (rationMisalignedChains / numberOfInputCoords > thrRatioMahalanobis)
+	{
+		globalAlignment = false;
+	}
+
+	std::cout << "------> Global alignment score: " << rationMisalignedChains / numberOfInputCoords << std::endl;
+	
+	// Local alignment analysis
+	std::cout << "---------------- Local misaligment analysis" << std::endl;
+
+	// Use votting as criteria for local misalignment detection
+	if (voteCriteria)
+	{
+		for (size_t n = 0; n < nSize; n++)
+		{
+			std::vector<resMod> resMod_image;
+			getResModByImage(n, resMod_image);
+
+			size_t numberResMod = resMod_image.size();
+
+			if (numberResMod > 1)
+			{
+				double rationMisalignedResid = 0.0;
+
+				for (size_t i = 0; i < numberResMod; i++)
+				{
+					if (resMod_image[i].mahalanobisDistance > 1)
+					{
+						rationMisalignedResid += 1;
+					}
+				}
+
+				rationMisalignedResid /= numberResMod;
+
+				if (rationMisalignedResid > thrRatioMahalanobis)
+				{
+					localAlignment[n] = false;
+					std::cout << "------> Local misalignment detected at image: " << n << " with ratio " << rationMisalignedResid << std::endl;
+				}
+			}
+
+			else
+			{
+				std::cout << "ERROR: impossible to study misalignment in tilt-image " << n << ". Number of residuals for this image: " << numberResMod << std::endl;
+				localAlignment[n] = false;
+			}
+		}
+	}
+	else // Use average as criteria for local misalignment detection
+	{
+		for (size_t n = 0; n < nSize; n++)
+		{
+			std::vector<resMod> resMod_image;
+			getResModByImage(n, resMod_image);
+
+			size_t numberResMod = resMod_image.size();
+
+			if (numberResMod > 1)
+			{
+				sumMahaDist = 0;
+				sumMahaDist2 = 0;
+
+				for (size_t i = 0; i < numberResMod; i++)
+				{
+					if (removeOutliers && resMod_image[i].mahalanobisDistance < mahaThr)
+					{
+						sumMahaDist += resMod_image[i].mahalanobisDistance;
+						sumMahaDist2 += resMod_image[i].mahalanobisDistance * resMod_image[i].mahalanobisDistance;
```

Comment:
Maybe put `resMod_image[i].mahalanobisDistance` in a variable and use it to avoid repeating it 3 times

Reply:
fixed